src/SourceGenerators/InstrumentationDefinitionsGenerator.cs

```diff
@@ -0,0 +1,280 @@
+// <copyright file="InstrumentationDefinitionsGenerator.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace SourceGenerators;
+
+[Generator]
+public class InstrumentationDefinitionsGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var classesWithAttributes = context.SyntaxProvider.CreateSyntaxProvider(
+                static (node, _) => IsAttributedClass(node),
+                static (node, _) => GetClassesMarkedByInstrumentMethodAttribute(node))
+            .Where(static m => m != null);
+
+        IncrementalValueProvider<(Compilation, ImmutableArray<ClassDeclarationSyntax?>)> compilationAndClasses
+            = context.CompilationProvider.Combine(classesWithAttributes.Collect());
+
+        context.RegisterSourceOutput(
+            compilationAndClasses,
+            static (context, source) => Generate(source.Item1, source.Item2, context));
+    }
+
+    private static void Generate(Compilation compilation, ImmutableArray<ClassDeclarationSyntax?> classesWithAttributes, SourceProductionContext context)
+    {
+        if (classesWithAttributes.IsDefaultOrEmpty)
+        {
+            return;
```

Comment:
hum ... should a warning be issued in this case? It seems valid to build everything without bytecode instrumentations, but, it seems atypical...

Reply:
It is just fast handle scenario taken from linked example. I do not think it will harm us, also adding warnings is not what I would like to do becasue adding bytecode instrumentation is not typical task for us. If new instrumentation will not occur locally we should detect it on